src/algorithms/postburn/PostBurn.cc

```diff
@@ -0,0 +1,153 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2024 Alex Jentsch, Jihee Kim, Brian Page
+//
+
+#include "PostBurn.h"
+
+#include <DD4hep/Alignments.h>
+#include <DD4hep/DetElement.h>
+#include <DD4hep/Objects.h>
+#include <DD4hep/VolumeManager.h>
+#include <Evaluator/DD4hepUnits.h>
+#include <Math/GenVector/Cartesian3D.h>
+#include <Math/GenVector/DisplacementVector3D.h>
+#include <edm4hep/Vector3f.h>
+#include <edm4hep/utils/vector_utils.h>
+#include <cmath>
+#include <gsl/pointers>
+#include <vector>
+#include <TVector3.h>
+#include <TLorentzVector.h>
+
+#include "algorithms/postburn/PostBurnConfig.h"
+
+void eicrecon::PostBurn::init(std::shared_ptr<spdlog::logger> &logger) {
+    m_log       = logger;
+}
+
+void eicrecon::PostBurn::process(
+    const PostBurn::Input& input,
+    const PostBurn::Output& output) const {
+
+    const auto [mcparts, recparticles, recParticlesAssoc] = input;
+    auto [outputParticles] = output;
+
+    bool      pidAssumePionMass = m_cfg.pidAssumePionMass;
+    double    crossingAngle    = m_cfg.crossingAngle;
+    double    pidPurity        = m_cfg.pidPurity;
+    bool      correctBeamFX    = m_cfg.correctBeamFX;
+    bool      pidUseMCTruth    = m_cfg.pidUseMCTruth;
+
+    bool      hasBeamHadron    = false;
+    bool      hasBeamLepton    = false;
+
+    //read MCParticles information for status == 1 particles and post-burn
+
+    ROOT::Math::PxPyPzEVector  e_beam(0.,0.,0.,0.);
+    ROOT::Math::PxPyPzEVector  h_beam(0.,0.,0.,0.);
+
+    // First, extract beams, flag decides if using beam kinematics as-stored, or only the crossing-angle boost/rotation
+    // This functionality is important for functionality for reconstructed tracks
+    if(correctBeamFX == true){
+        for (const auto& p: *mcparts) {
+
+            if(p.getGeneratorStatus() == 4 && (p.getPDG() == 2212 || p.getPDG() == 2112)) { //look for "beam" proton/neutron
+                h_beam.SetPxPyPzE(p.getMomentum().x, p.getMomentum().y, p.getMomentum().z, p.getEnergy());
+                hasBeamHadron = true;
+            }
+            if(p.getGeneratorStatus() == 4 && p.getPDG() == 11) { //look for "beam" electron
+                e_beam.SetPxPyPzE(p.getMomentum().x, p.getMomentum().y, p.getMomentum().z, p.getEnergy());
+                hasBeamLepton = true;
+            }
+        }
```

Comment:
Doesn't this add more unnecessary overhead and abstraction for something which is performing a simple operation?

Reply:
No, there is no overhead to this. It centralizes repeated functionality to prevent fragmentation in approaches. As soon as someone starts using He4 beams, we then only need to change one place instead of 15 places.

Reply:
The better(?) alternative would be using the centralized collections MCBeamElectrons and MCBeamProtons (which could be adapted to include neutrons) so the search is only conducted once globally.

Reply:
Agreed as well. Do we have those now?

Reply:
If we have to adapt something to use it for this PR, doesn't it make more sense to just use the functions from "beams.h"? This is a really simple portion of the code, it can be easily adjusted to make it nicer later.

Reply:
This was something that had to be addressed to match the beam.h implementation anyway: https://github.com/eic/EICrecon/pull/1497/commits/86f79099a2df9536a8bc5eb59ed9beee34f73ad4

Reply:
I am using the beam.h implementation for now - can fix on the next PR when we get the postburner working for reco branches. Trying to push my commits now.